C side

caiovini980 opened the pull request.

caiovini980 committed:
Merge branch 'develop' into c-side

caiovini980 committed:
feat: implemented projectile collision for player and enemies

Nevasca commented on Game/space-invaders/src/enemies/EnemyManager.cpp:
```diff
@@ -137,6 +139,8 @@ void EnemyManager::SpawnEnemies(const LevelDefinition& level)
             position.x += (level.Padding + m_EnemySize.x) * x;
             position.y += (level.Padding + m_EnemySize.y) * y;
+
+            // GameObject newEnemy = 
```
Please remove hanging commented code

Nevasca commented on Game/space-invaders/src/enemies/EnemyManager.cpp:
```diff
@@ -148,6 +152,19 @@ void EnemyManager::SpawnEnemies(const LevelDefinition& level)
     }
 }
 
+void EnemyManager::CheckCollisions(GameObject& projectile, GameObject& hitObject)
```
I belive it will be better for the Game to check the collision and then call a HandleEnemyHit or HandlePlayerHit after confirming it has collided. There will be less instruction jump for checking if collided or not

Nevasca commented on Game/space-invaders/src/enemies/EnemyManager.h:
```diff
@@ -1,5 +1,6 @@
 #pragma once
 #include <vector>
+#include <algorithm>
```
Do we really need this include? Not sure where it's used on EnemyManager

Nevasca commented on Game/space-invaders/src/Game.cpp:
```diff
@@ -98,6 +101,12 @@ void Game::UpdatePlayerProjectiles(float deltaTime)
+        // check collision with enemies or block
+        for (auto& enemy : m_Level->GetEnemies())
+        {
+            m_PlayerManager->CheckCollisions(projectile, enemy);
```
Make the collision check here and call HandlePlayerHit only if hit

Nevasca commented on Game/space-invaders/src/Game.cpp:
```diff
@@ -111,6 +120,10 @@ void Game::UpdateEnemyProjectiles(float deltaTime)
+
+        GameObject& player = m_PlayerManager->GetPlayer();
+        
+        m_Level->CheckCollisionWithEnemies(projectile, player);
```
Same here, check for collision on Game and call m_Level->HandleEmenyHit only if collided

Nevasca commented on Game/space-invaders/src/GameLevel.h:
```diff
@@ -15,6 +15,9 @@ class GameLevel
     void Update(float deltaTime);
     void Render(const SpriteRenderer& renderer);
 
+    std::vector<GameObject>& GetEnemies() const { return m_EnemyManager->GetEnemies(); }
```
Add inline keyword here

Nevasca commented on Game/space-invaders/src/enemies/EnemyManager.h:
```diff
@@ -15,7 +16,10 @@ class EnemyManager
     void Render(const SpriteRenderer& renderer);
+    void CheckCollisions(GameObject& projectile, GameObject& hitObject);
 
+    std::vector<GameObject>& GetEnemies() { return m_Enemies; }
```
Add inline keyword here

Nevasca commented on Game/space-invaders/src/GameObject.h:
```diff
@@ -17,6 +17,7 @@ class GameObject
     glm::vec3 Color{1.f};
     float Rotation{0.f};
     bool Destroyed{false};
+    bool bIsPlayer{false};
```
We shouldn't add this flag here to keep GameObject as generic as possible

Nevasca commented on Game/space-invaders/src/players/PlayerManager.h:
```diff
@@ -19,6 +20,9 @@ class PlayerManager
     void Render(const SpriteRenderer& renderer) const;
+    void CheckCollisions(GameObject& projectile, GameObject& enemy);
+
+    GameObject& GetPlayer() const { return *m_Player; }
```
Add inline keyword here

caiovini980 committed:
refactor: changed the way collisions are being made

before the player check collision on the enemies, and vice versa. Now is the opposite

caiovini980 committed:
fix and feat: implemented changes from code review and player lives

Nevasca commented on Game/space-invaders/src/enemies/EnemyManager.cpp:
```diff
@@ -2,6 +2,7 @@
 
 #include <iostream>
 
+#include "Collision.h"
```
Remove unused Collision include

Nevasca commented on Game/space-invaders/src/enemies/EnemyManager.cpp:
```diff
@@ -148,6 +150,11 @@ void EnemyManager::SpawnEnemies(const LevelDefinition& level)
     }
 }
 
+void EnemyManager::CheckCollisions(GameObject& projectile)
```
Remove unused CheckCollisions method, from cpp and header

Nevasca commented on Game/space-invaders/src/players/PlayerManager.cpp:
```diff
@@ -2,6 +2,7 @@
 
 #include <GLFW/glfw3.h>
 
+#include "Collision.h"
```
Removed unused Collision include

Nevasca commented on Game/space-invaders/src/players/PlayerManager.h:
```diff
@@ -3,10 +3,11 @@
 
 #include <iostream>
 
+// #include "Game.h"
```
Remove hanging commented code

Nevasca commented on Game/space-invaders/src/Game.cpp:
```diff
@@ -111,6 +116,26 @@ void Game::UpdateEnemyProjectiles(float deltaTime)
+    for (auto& enemy : m_Level->GetEnemies())
+    {
+        if (!Collision::IsColliding(projectile, enemy)) { continue; }
+        if (enemy.Destroyed) { continue; }
```
Check for destroyed before checking for collision

Nevasca commented on Game/space-invaders/src/Game.cpp:
```diff
@@ -111,6 +116,26 @@ void Game::UpdateEnemyProjectiles(float deltaTime)
+
+        GameObject& player = m_PlayerManager->GetPlayer();
+        if (!Collision::IsColliding(projectile, player)) { return; }
+        if (player.Destroyed) { return; }
```
Check for destroyed before checking for collision

caiovini980 committed:
refactor: removed unnused includes, comments and changed order of checking collisions on Game.cpp
